File: sdk/core/android/src/main/kotlin/network/NetworkWarmer.kt
package com.rivalapexmediation.sdk.network

import android.os.Build
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import okhttp3.ConnectionPool
import okhttp3.OkHttpClient
import okhttp3.Protocol
import okhttp3.Request
import java.net.InetAddress
import java.net.URL
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

/**
 * NetworkWarmer provides HTTP preconnect and DNS prefetch functionality
 * to reduce cold-start latency for auction requests.
 * 
 * Key features:
 * - DNS prefetch on SDK init (caches DNS resolution)
 * - HTTP/2 connection preconnect (establishes TLS + ALPN early)
 * - Connection pooling warmup
 * - Thread-safe singleton pattern
 * 
 * Usage:
 * ```kotlin
 * // At SDK init time:
 * NetworkWarmer.warmup("https://api.apexmediation.com")
 * 
 * // Get pre-warmed client for requests:
 * val client = NetworkWarmer.getWarmedClient()
 * ```
 */
object NetworkWarmer {
    
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private val isWarmedUp = AtomicBoolean(false)
    private val dnsCache = ConcurrentHashMap<String, List<InetAddress>>()
    private var warmedEndpoints = mutableSetOf<String>()
    
    // Shared client with optimal connection pooling
    private val sharedClient: OkHttpClient by lazy {
        OkHttpClient.Builder()
            .connectionPool(ConnectionPool(
                maxIdleConnections = Runtime.getRuntime().availableProcessors().coerceIn(2, 8),
                keepAliveDuration = 5,
                timeUnit = TimeUnit.MINUTES
            ))
            .connectTimeout(2, TimeUnit.SECONDS)
            .readTimeout(5, TimeUnit.SECONDS)
            .writeTimeout(5, TimeUnit.SECONDS)
            .callTimeout(6, TimeUnit.SECONDS)
            .retryOnConnectionFailure(false)
            .followRedirects(false)
            .followSslRedirects(false)
            .protocols(listOf(Protocol.HTTP_2, Protocol.HTTP_1_1))
            // DNS cache integration
            .dns { hostname ->
                dnsCache[hostname] ?: InetAddress.getAllByName(hostname).toList().also {
                    dnsCache[hostname] = it
                }
            }
            .build()
    }
    
    /**
     * Warm up connections to the specified endpoint.
     * This performs:
     * 1. DNS prefetch (resolves hostname to IP addresses)
     * 2. HTTP preconnect (establishes TCP + TLS connection)
     * 
     * Call this at SDK initialization time for best cold-start performance.
     * 
     * @param endpoint Base URL of the auction endpoint (e.g., "https://api.apexmediation.com")
     */
    fun warmup(endpoint: String) {
        if (warmedEndpoints.contains(endpoint)) return
        
        scope.launch {
            try {
                val url = URL(endpoint)
                val hostname = url.host
                
                // Step 1: DNS prefetch
                prefetchDns(hostname)
                
                // Step 2: HTTP preconnect (HEAD request to warm connection pool)
                preconnect(endpoint)
                
                warmedEndpoints.add(endpoint)
                isWarmedUp.set(true)
            } catch (e: Exception) {
                // Warmup is best-effort; don't crash on failure
            }
        }
    }
    
    /**
     * Prefetch DNS for a hostname, caching the result.
     */
    private suspend fun prefetchDns(hostname: String) {
        if (dnsCache.containsKey(hostname)) return
        
        try {
            val addresses = InetAddress.getAllByName(hostname).toList()
            if (addresses.isNotEmpty()) {
                dnsCache[hostname] = addresses
            }
        } catch (e: Exception) {
            // DNS resolution failed; will retry on actual request
        }
    }
    
    /**
     * Establish a connection to the endpoint using a lightweight HEAD request.
     * This warms the connection pool for subsequent requests.
     */
    private fun preconnect(endpoint: String) {
        try {
            val request = Request.Builder()
                .url("$endpoint/health")
                .head()
                .build()
            
            // Execute synchronously but with short timeout
            sharedClient.newBuilder()
                .callTimeout(2, TimeUnit.SECONDS)
                .build()
                .newCall(request)
                .execute()
                .close()
        } catch (e: Exception) {
            // Preconnect is best-effort
        }
    }
    
    /**
     * Get the pre-warmed OkHttpClient instance.
     * This client has DNS caching and connection pooling already configured.
     */
    fun getWarmedClient(): OkHttpClient = sharedClient
    
    /**
     * Check if warmup has completed for at least one endpoint.
     */
    fun isWarmedUp(): Boolean = isWarmedUp.get()
    
    /**
     * Get cached DNS entries for a hostname, if available.
     */
    fun getCachedDns(hostname: String): List<InetAddress>? = dnsCache[hostname]
    
    /**
     * Clear all cached DNS entries and warmed endpoints.
     * Primarily for testing.
     */
    fun reset() {
        dnsCache.clear()
        warmedEndpoints.clear()
        isWarmedUp.set(false)
    }
    
    /**
     * Get diagnostics about current warmup state.
     */
    fun getDiagnostics(): Map<String, Any> = mapOf(
        "isWarmedUp" to isWarmedUp.get(),
        "warmedEndpoints" to warmedEndpoints.toList(),
        "dnsCacheSize" to dnsCache.size,
        "cachedHosts" to dnsCache.keys.toList(),
        "connectionPoolSize" to Runtime.getRuntime().availableProcessors().coerceIn(2, 8),
        "sdkVersion" to (try { com.rivalapexmediation.sdk.BuildConfig.SDK_VERSION } catch (_: Throwable) { "unknown" })
    )
}

File: sdk/core/unity/Runtime/Core/NetworkWarmer.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Net;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using UnityEngine;

namespace ApexMediation.Core
{
    /// <summary>
    /// NetworkWarmer provides HTTP preconnect and DNS prefetch functionality
    /// to reduce cold-start latency for auction requests.
    /// 
    /// Key features:
    /// - DNS prefetch on SDK init (caches DNS resolution)
    /// - HTTP preconnect (establishes TLS early)
    /// - Connection pooling warmup
    /// - Thread-safe singleton pattern
    /// </summary>
    public sealed class NetworkWarmer
    {
        private static readonly Lazy<NetworkWarmer> _instance = 
            new Lazy<NetworkWarmer>(() => new NetworkWarmer());
        
        public static NetworkWarmer Instance => _instance.Value;
        
        private readonly ConcurrentDictionary<string, IPAddress[]> _dnsCache = 
            new ConcurrentDictionary<string, IPAddress[]>();
        private readonly HashSet<string> _warmedEndpoints = new HashSet<string>();
        private readonly object _lock = new object();
        private volatile bool _isWarmedUp;
        
        // Shared HttpClient with connection pooling
        private readonly HttpClient _httpClient;
        
        private NetworkWarmer()
        {
            var handler = new HttpClientHandler
            {
                MaxConnectionsPerServer = Math.Max(2, Math.Min(Environment.ProcessorCount, 8)),
                AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate
            };
            
            _httpClient = new HttpClient(handler)
            {
                Timeout = TimeSpan.FromSeconds(5)
            };
        }
        
        /// <summary>
        /// Whether at least one endpoint has been warmed up.
        /// </summary>
        public bool IsWarmedUp => _isWarmedUp;
        
        /// <summary>
        /// Warm up connections to the specified endpoint.
        /// This performs:
        /// 1. DNS prefetch (resolves hostname to IP addresses)
        /// 2. HTTP preconnect (establishes TCP + TLS connection)
        /// </summary>
        /// <param name="endpoint">Base URL of the auction endpoint</param>
        public void Warmup(string endpoint)
        {
            lock (_lock)
            {
                if (_warmedEndpoints.Contains(endpoint))
                    return;
            }
            
            // Run warmup on thread pool to avoid blocking
            Task.Run(async () =>
            {
                try
                {
                    var uri = new Uri(endpoint);
                    var hostname = uri.Host;
                    
                    // Step 1: DNS prefetch
                    await PrefetchDnsAsync(hostname);
                    
                    // Step 2: HTTP preconnect
                    await PreconnectAsync(endpoint);
                    
                    lock (_lock)
                    {
                        _warmedEndpoints.Add(endpoint);
                        _isWarmedUp = true;
                    }
                }
                catch (Exception ex)
                {
                    // Warmup is best-effort
                    Debug.Log($"[ApexMediation] NetworkWarmer warmup failed: {ex.Message}");
                }
            });
        }
        
        /// <summary>
        /// Warm up connections asynchronously.
        /// </summary>
        public async Task WarmupAsync(string endpoint)
        {
            lock (_lock)
            {
                if (_warmedEndpoints.Contains(endpoint))
                    return;
            }
            
            try
            {
                var uri = new Uri(endpoint);
                var hostname = uri.Host;
                
                await PrefetchDnsAsync(hostname);
                await PreconnectAsync(endpoint);
                
                lock (_lock)
                {
                    _warmedEndpoints.Add(endpoint);
                    _isWarmedUp = true;
                }
            }
            catch (Exception ex)
            {
                Debug.Log($"[ApexMediation] NetworkWarmer warmup failed: {ex.Message}");
            }
        }
        
        private async Task PrefetchDnsAsync(string hostname)
        {
            if (_dnsCache.ContainsKey(hostname))
                return;
            
            try
            {
                var addresses = await Dns.GetHostAddressesAsync(hostname);
                if (addresses.Length > 0)
                {
                    _dnsCache[hostname] = addresses;
                }
            }
            catch
            {
                // DNS resolution failed; will retry on actual request
            }
        }
        
        private async Task PreconnectAsync(string endpoint)
        {
            try
            {
                using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(2));
                using var request = new HttpRequestMessage(HttpMethod.Head, endpoint + "/health");
                
                var response = await _httpClient.SendAsync(request, cts.Token);
                response.Dispose();
            }
            catch
            {
                // Preconnect is best-effort
            }
        }
        
        /// <summary>
        /// Get cached DNS entries for a hostname, if available.
        /// </summary>
        public IPAddress[] GetCachedDns(string hostname)
        {
            return _dnsCache.TryGetValue(hostname, out var addresses) ? addresses : null;
        }
        
        /// <summary>
        /// Get the shared HttpClient with pre-warmed connections.
        /// </summary>
        public HttpClient GetWarmedClient() => _httpClient;
        
        /// <summary>
        /// Clear all cached DNS entries and warmed endpoints.
        /// Primarily for testing.
        /// </summary>
        public void Reset()
        {
            lock (_lock)
            {
                _dnsCache.Clear();
                _warmedEndpoints.Clear();
                _isWarmedUp = false;
            }
        }
        
        /// <summary>
        /// Get diagnostics about current warmup state.
        /// </summary>
        public Dictionary<string, object> GetDiagnostics()
        {
            lock (_lock)
            {
                return new Dictionary<string, object>
                {
                    { "isWarmedUp", _isWarmedUp },
                    { "warmedEndpoints", new List<string>(_warmedEndpoints) },
                    { "dnsCacheSize", _dnsCache.Count },
                    { "cachedHosts", new List<string>(_dnsCache.Keys) },
                    { "connectionPoolSize", Math.Max(2, Math.Min(Environment.ProcessorCount, 8)) }
                };
            }
        }
    }
}
